libcxx/test/std/utilities/function.objects/func.bind.partial/bind_front.nttp.pass.cpp

```diff
@@ -0,0 +1,355 @@
+      CopyMoveInfo info;
+      auto f = std::bind_front<was_moved>(info);
+      assert(std::move(f)());
+    }
+  }
+}
+
+constexpr void test_perfect_forwarding_call_wrapper() {
+  { // Make sure we call the correctly cv-ref qualified operator()
+    // based on the value category of the bind_front<NTTP> unspecified-type.
+    struct X {
+      constexpr int operator()() & { return 1; }
+      constexpr int operator()() const& { return 2; }
+      constexpr int operator()() && { return 3; }
+      constexpr int operator()() const&& { return 4; }
+    };
+
+    auto f  = std::bind_front<X{}>();
+    using F = decltype(f);
+    assert(static_cast<F&>(f)() == 2);
+    assert(static_cast<const F&>(f)() == 2);
+    assert(static_cast<F&&>(f)() == 2);
+    assert(static_cast<const F&&>(f)() == 2);
+  }
+
+  // Call to `bind_front<NTTP>` unspecified-type's operator() should always result in call to the const& overload of the underlying function object.
+  {
+    { // Make sure unspecified-type is still callable when we delete the & overload.
+      struct X {
+        int operator()() & = delete;
+        int operator()() const&;
+        int operator()() &&;
+        int operator()() const&&;
+      };
+
+      using F = decltype(std::bind_front<X{}>());
+      static_assert(std::invocable<F&>);
+      static_assert(std::invocable<const F&>);
+      static_assert(std::invocable<F>);
+      static_assert(std::invocable<const F>);
+    }
+
+    { // Make sure unspecified-type is not callable when we delete the const& overload.
+      struct X {
+        int operator()() &;
+        int operator()() const& = delete;
+        int operator()() &&;
+        int operator()() const&&;
+      };
+
+      using F = decltype(std::bind_front<X{}>());
+      static_assert(!std::invocable<F&>);
+      static_assert(!std::invocable<const F&>);
+      static_assert(!std::invocable<F>);
+      static_assert(!std::invocable<const F>);
+    }
+
+    { // Make sure unspecified-type is still callable when we delete the && overload.
+      struct X {
+        int operator()() &;
+        int operator()() const&;
+        int operator()() && = delete;
+        int operator()() const&&;
+      };
+
+      using F = decltype(std::bind_front<X{}>());
+      static_assert(std::invocable<F&>);
+      static_assert(std::invocable<const F&>);
+      static_assert(std::invocable<F>);
+      static_assert(std::invocable<const F>);
+    }
+
+    { // Make sure unspecified-type is still callable when we delete the const&& overload.
+      struct X {
+        int operator()() &;
+        int operator()() const&;
+        int operator()() &&;
+        int operator()() const&& = delete;
+      };
+
+      using F = decltype(std::bind_front<X{}>());
+      static_assert(std::invocable<F&>);
+      static_assert(std::invocable<const F&>);
+      static_assert(std::invocable<F>);
+      static_assert(std::invocable<const F>);
+    }
+  }
+
+  { // Test perfect forwarding
```

Comment:
I think in the test suite what is missing and important is that the bond arguments are correct perfectly forward-like  as the returned object. given
```
constexpr struct F {
 
  int operator()(int&, double) const;
  int operator()(const int&, double) const;
  int operator()(int&&, double) const;
  int operator()(const int&&, double) const;
} f{};
```
and

```
auto r = bind_front<f>(5);
```

the results of
```
r(1.0);
std::as_const(r)(1.0);
std::move(r)(1.0);
std::move(std::as_const(r))(1.0);
```

should call different overloads

Reply:
Done: https://github.com/llvm/llvm-project/pull/165096/commits/4689f86dfe437a356fd296328783ed5c9c276fe6